master/buildbot/www/authz/authz.py

```diff
@@ -89,10 +89,13 @@ def assertUserAllowed(self, ep, action, options, userDetails):
                     if owner is not None:
                         roles.update(
                             self.getOwnerRolesFromUser(userDetails, owner))
-                if rule.role not in roles:
-                    if rule.defaultDeny:
-                        raise Forbidden(
-                            "you need to have role '%s'" % (rule.role,))
-                elif not rule.defaultDeny:
-                    defer.returnValue(None)
+                for role in roles:
+                    if self.match(role, rule.role):
```

Comment:
I think that there is a problem with authz testing. Please correct me if I'm wrong.

As Authz.assertUserAllowed method does not return which combination of rule and role succeeded
 the check we can't test that exact one rule succeeded but not some other.
For example if I want to check regex-based role check I need to add some rule with role pattern.
How can I be shure that exactly that rule succeeded or failed in test but not some other in
long list of rules?

I think that there are two solutions:
1) return combination of rule and role that succeeded the authz check
(it seems that return value is not checked anywhere in buildbot).
So we can compare it in test and make tests more reliable.
2) override list of rules and list of roles in test method to make tests more isolated and predictable.

Maybe that's why where is a bug when idefaultDeny=False gives access if access should not be given